Add fs_find_file function.

zbsz opened the pull request:
Added file listing which can be later used to load apps.

zbsz commented:
Changed to only store some flags per page instead of having a list of files.
I'm not entirely sure about some parts here, especially `fs_find_file` return value in case the file is not found is not perfect.

jwise commented on rcore/flash.c:
```diff
@@ -117,8 +117,16 @@ struct page_hdr {
 
-static void _fs_read_page_hdr(int pg, struct page_hdr *p) {
-    flash_read_bytes(REGION_FS_START + pg * REGION_FS_PAGE_SIZE, (uint8_t *)p, sizeof(struct page_hdr));
+#define MAX_FILENAME_LEN 64
```
Where'd this constant come from?  Is that actually the maximum?

zbsz replied:
just my assumption, will comment and add check in `init`

jwise commented on rcore/flash.c:
```diff
@@ -286,21 +289,20 @@ uint32_t flash_get_resource_address(uint16_t slot_id)
 
 const File fs_find_file(const char *name)
```
maybe have this populate into a pointer passed in?  signature something like:

`int fs_open(struct file *file, const char *name);`

zbsz replied:
yup, that would make result explicit

jwise commented on rcore/flash.c:
```diff
@@ -117,34 +117,56 @@ struct page_hdr {
+    PageState state : 2;
+} PageFlags;
+
+static PageFlags page_flags[REGION_FS_N_PAGES];
```
I think local style is lowercase; `struct pageflags` and `typedef enum pagestate { ... } pagestate_t;`

jwise commented on rcore/flash.c:
```diff
@@ -117,34 +117,56 @@ struct page_hdr {
+    PageState state : 2;
+} PageFlags;
+
+static PageFlags page_flags[REGION_FS_N_PAGES];
```
also, is the compiler smart enough to bitpack that struct?  (not critical for this commit)

zbsz replied:
Didn't think about it, it actually is not able to pack it. Any suggestion how to fix it? What should I use instead of this struct?

jwise commented on rcore/flash.c:
```diff
@@ -158,54 +180,55 @@ static void _fs_init()
             KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d deletion not complete; I can't deal with this; go boot PebbleOS to clean up first", pg);
             _fs_valid = 0;
             return;
         }
```
do an integrity check if MAX_FILENAME_LEN doesn't come from something in PebbleOS?

jwise commented:
This looks really good.  I would just merge it and fix my complaints myself, though I won't have time tonight.  If you get a chance maybe take a look at those?

jwise commented on rcore/flash.c:
```diff
@@ -260,6 +283,30 @@ uint32_t flash_get_resource_address(uint16_t slot_id)
 }
+
+const File fs_find_file(const char *name)
+{
```
make sure to respect `_fs_valid` -- we haven't finished populating our internal state otherwise

zbsz commented:
Addressed review comments.

Changed page flags to use basic array and bit operations instead of struct (which wasn't packed). Let me know if there is better way to implement this (maybe more clever bit operations).

Should I also change style of `page_state` enum values? Kept camel case for now.

zbsz commented on rcore/flash.c:
```diff
@@ -158,54 +189,58 @@ static void _fs_init()
-            }
+
+        if (hdr->filename_len > MAX_FILENAME_LEN)
+            KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d has unexpectedly long file name: %d; it may cause further errors", pg, hdr->filename_len);
```
only loggin here, not setting to invalid, as this is not FS error, only wrong assumptions on my side, this will not crash, such file will just not be found by `fs_find_file` 

I'm no longer sure if this optimization with reading a name in one go is worth it, considering this possible bug. Should I remove that and just read filename separately (would require allocation here)?

jwise commented on rcore/flash.c:
```diff
@@ -117,34 +117,65 @@ struct page_hdr {
+
+static uint8_t _fs_page_flags[(REGION_FS_N_PAGES + 3) >> 2];
+
+static void _fs_set_page_state(uint16_t pg, enum page_state state)
```
Yeah, this is how I'd do it.  Looks good to me.

jwise commented:
Looks good to me.  The problem with having `MAX_FILENAME_LEN` be 256 bytes is that it uses up a lot of stack memory -- and using `alloca()` doesn't solve it, since you need the worst case, anyway.  So I think it's ok with me that `MAX_FILENAME_LEN` is 64 bytes.  (Maybe even 32 would be ok.)

This is great work!  From here we should be able to load an `appdb`, and then should be able to also more reliably load watchfaces / watchapps from their correct locations on flash.
